Add structure for pluggable hash libraries

hickeyma opened the pull request:
Needs to handle vLLM hash algorithm support because vLLM  allows hashes in addition to tokens for querying LMCache.

Fix #1079

hickeyma committed:
Add xxhash to token database

Set xxhash as hashing library as first stage to run end to end

Signed-off-by: Martin Hickey <martin.hickey@ie.ibm.com>

hickeyma commented:
@YaoJiayi This is a starting point for discussion on how best to implement this. Stared with hard coding `xxhash` library for now in `TokenDatabase`.

YaoJiayi reviewed (commented):
Thanks @hickeyma ! We need to use the python internal `hash()` for now since that's what vllm does. Can we add a config to support different hashing algo?

hickeyma commented:
> Thanks @hickeyma ! We need to use the python internal hash() for now since that's what vllm does. Can we add a config to support different hashing algo?

@YaoJiayi I'll investigate this further.

hickeyma committed:
Add hashing functions supported in vLLM

Following discussion with Jiayi Yao, LMCache needs
to be able to support vLLM hashing capability as
vLLM will use hashes instead of tokens in some scenarios.

Signed-off-by: Martin Hickey <martin.hickey@ie.ibm.com>

hickeyma committed:
Fix cache controller example

Fix config file `example.yaml` which was missing worker
port property and contained old worker url property.

Signed-off-by: Martin Hickey <martin.hickey@ie.ibm.com>

hickeyma committed:
Fix SegmentTokenDatabase init of hash function

Signed-off-by: Martin Hickey <martin.hickey@ie.ibm.com>

hickeyma committed:
Remove developer tracing

Signed-off-by: Martin Hickey <martin.hickey@ie.ibm.com>

hickeyma committed:
Add common hash function initialization to base TokenDatabase object

The initialization of the hash function is duplicated across both
derived objects of TokenDatabase. It therefore makes more sense to
add it to one place in TokenDatabase.

Signed-off-by: Martin Hickey <martin.hickey@ie.ibm.com>

hickeyma commented:
@YaoJiayi PR is ready for review now. Is it ok to just support the vLLM hashing functions for now or do you want me to wrap `xxhash` in this PR as well, as per @xiaguan comment in #1079?

YaoJiayi commented on examples/cache_controller/example.yaml:
```diff
@@ -6,4 +6,4 @@ max_local_cpu_size: 5
 controller_url: "localhost:9001"
-lmcache_worker_url: "localhost:8001"
\ No newline at end of file
+lmcache_worker_port: 8001
```
Good catch!

YaoJiayi reviewed (approved):
LGTM!
